exist-core/src/main/java/org/exist/xquery/functions/fn/FunXmlToJson.java

```diff
@@ -0,0 +1,244 @@
+                                jsonGenerator.writeEndArray();
+                                break;
+                            case "boolean":
+                                tempBoolean = !("".equals(tempString) || "0".equals(tempString) || "false".equals(tempString));
+                                jsonGenerator.writeBoolean(tempBoolean);
+                                break;
+                            case "map":
+                                while (mapkeyArrayList.size() > 0 && mapkeyArrayList.remove(mapkeyArrayList.size() - 1) != stackSeparator) {
+                                }
+                                jsonGenerator.writeEndObject();
+                                break;
+                            case "null":
+                                if (tempStringBuilder.length() != 0) {
+                                    throw new XPathException(ErrorCodes.FOJS0006, "Invalid XML representation of JSON. Found non-empty XML null element.");
+                                }
+                                jsonGenerator.writeNull();
+                                break;
+                            case "number":
+                                double tempDouble = Double.parseDouble(tempString);
+                                jsonGenerator.writeNumber(tempDouble);
+                                break;
+                            case "string":
+                                if (elementValueIsEscaped == true) {
+                                    //TODO: any unescaped occurrence of quotation mark, backspace, form-feed, newline, carriage return, tab, or solidus is replaced by \", \b, \f, \n, \r, \t, or \/ respectively;
+                                    //TODO: any other codepoint in the range 1-31 or 127-159 is replaced by an escape in the form <backslash>uHHHH where HHHH is the upper-case hexadecimal representation of the codepoint value.
+                                    jsonGenerator.writeString(unescapeEscapedJsonString(tempString));
+                                } else {
+                                    //TODO: any other codepoint in the range 1-31 or 127-159 is replaced by an escape in the form <backslash>uHHHH where HHHH is the upper-case hexadecimal representation of the codepoint value.
+                                    jsonGenerator.writeString(tempString);
+                                }
+                                break;
+                            default:
+                                throw new XPathException(ErrorCodes.FOJS0006, "Invalid XML representation of JSON. Found XML element which is not one of [map, array, null, boolean, number, string].");
+                        }
+                    default:
+                        break;
+                }
+            }
+        } catch (JsonGenerationException e) {
+            throw new XPathException(ErrorCodes.FOJS0006, "Invalid XML representation of JSON.");
+        } catch (XMLStreamException e) {
```

Comment:
line 202 and 204 can be combined, what about changing the `"IOException"` string into `e.getMessage()` ?

Reply:
Sounds good.